[ClangImporter] Make conformance loading lazier.

jrose-apple opened the pull request:
Previously, the importer queued up conformances to complete once it was done importing the current batch of declarations. However, if there was a serialized Swift module that extended an imported type to add a conformance in exactly the wrong way, the importer could end up asking for that conformance later—even before the reference to the imported type was resolved. This led to a crash in the deserializer "while reading conformance for type X".

Instead of this "pending actions" queue, we can just use the mechanisms already in place for lazily loading conformances. That way they'll get filled out on demand, which is better all around anyway. This does mean putting the requirement signature into the "lazy" part of the conformance, though.

This does as a side effect mean that *all* of the witnesses for the imported conformance may be opaque—that is, they will never be devirtualized to a particular implementation. However, they previously would have referred to methods implemented in Objective-C anyway, which are always dispatched with objc_msgSend. So this should have no practical effect.

There's also a second commit that makes deserialization of Swift conformances a little lazier as well, now that we have that ability. This isn't critical to this fix and we could easily leave it for later.

rdar://problem/32346184

jrose-apple committed:
[ClangImporter] Make conformance loading lazier.

Previously, the importer queued up conformances to complete once it
was done importing the current batch of declarations. However, if
there was a serialized Swift module that extended an imported type to
add a conformance in exactly the wrong way, the importer could end up
asking for that conformance later---even before the reference to the
imported type was resolved. This led to a crash in the deserializer
"while reading conformance for type X".

Instead of this "pending actions" queue, we can just use the
mechanisms already in place for lazily loading conformances. That way
they'll get filled out on demand, which is better all around anyway.
This does mean putting the requirement signature into the "lazy" part
of the conformance, though.

This does as a side effect mean that /all/ of the witnesses for the
imported conformance may be opaque---that is, they will never be
devirtualized to a particular implementation. However, they previously
would have referred to methods implemented in Objective-C anyway,
which are always dispatched with objc_msgSend. So this should have no
practical effect.

rdar://problem/32346184

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@DougGregor: after discussing this with @ematejska we decided the story with a flag didn't make enough sense—either we're comfortable with this for 4.0 or we're not. I'm personally a lot happier than I thought I would be because the lazy mechanism is already there; if I had had to add that for the first time I'd be a lot more apprehensive.

DougGregor reviewed (approved):
This was smaller than I expected, which is great. Thank you, LGTM.

jrose-apple commented:
Failing a number of internal tests. Stay tuned.

jrose-apple commented:
It looks like those are configuration issues on our end, not related to this pull request. Previous full-testing passed.

@swift-ci Please smoke test

jrose-apple committed:
[Serialization] Make requirement signature conformance loading lazy.

The previous commit needed this for the importer, so we might as well
take advantage of it in deserialization as well. No expected change in
behavior.

jrose-apple commented:
\*sigh\* Left in some conflict markers.

@swift-ci Please smoke test

jrose-apple commented:
@DougGregor, do you think we should take the serialization change for 4.0 as well, or just the importer work?

DougGregor reviewed (commented):
Thank you!

jrose-apple commented:
…wrong PR?
